src/main/java/com/selfxdsd/todos/Puzzles.java

```diff
@@ -22,74 +22,22 @@
  */
 package com.selfxdsd.todos;
 
-import com.selfxdsd.api.Project;
-import com.selfxdsd.api.Self;
-import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.http.ResponseEntity;
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.RestController;
-
-import java.io.IOException;
-
 /**
- * Puzzles REST Controller.
- * @author Mihai Andronache (amihaiemil@gmail.com)
+ * Representation of pdd puzzles. It also process the puzzles from
+ * an input.
+ * @author criske
  * @version $Id$
  * @since 0.0.1
+ * @param <Input> The type input used for processing.
  */
-@RestController
-public class Puzzles {
-
-    /**
-     * Self's core.
-     */
-    private final Self selfCore;
-
-    /**
-     * Puzzles through SSH.
-     */
-    private final SshPuzzles puzzles;
-
-    /**
-     * Ctor.
-     * @param selfCode Self Core, injected by Spring automatically.
-     * @param puzzles Puzzles through SSH.
-     */
-    @Autowired
-    public Puzzles(
-        final Self selfCode,
-        final SshPuzzles puzzles
-    ) {
-        this.selfCore = selfCode;
-        this.puzzles = puzzles;
-    }
+public interface Puzzles<Input> extends Iterable<Puzzle> {
 
     /**
-     * Trigger the reading/processing of the puzzles for the given Project.
-     * @param provider Provider name (github, gitlab etc).
-     * @param owner Owner login (user or organization name).
-     * @param name Simple name of the repository.
-     * @return Response OK.
-     * @throws IOException If something goes wrong.
+     * Process the puzzles from the given Input.
+     * These puzzles will be accessible through Iterator.
+     * @param input Input
+     * @throws PuzzlesProcessingException If something went wrong.
      */
-    @GetMapping(value = "/pdd/{provider}/{owner}/{name}")
-    public ResponseEntity<String> readPuzzles(
-        @PathVariable final String provider,
-        @PathVariable final String owner,
-        @PathVariable final String name
-    ) throws IOException {
-        final ResponseEntity<String> resp;
-        final Project project = this.selfCore.projects().getProjectById(
-            owner + "/" + name, provider
-        );
-        if(project == null) {
-            resp = ResponseEntity.badRequest().build();
-        } else {
-            resp = ResponseEntity.ok().build();
-            System.out.println(this.puzzles.read(project));
-        }
-        return resp;
-    }
+    void process(final Input input) throws PuzzlesProcessingException;
```

Comment:
@amihaiemil Maybe `process(<Input>)` could be in a different interface named like `PuzzlesProcessor<Intput>`?
So now, for example, DocumentPuzzles would look like
`DoumentPuzzles implements Puzzles, PuzzlesProcessor<String>`